### PR TITLE
Blueprinter: Render Performance Optimisations

### PR DESCRIPTION
**Pull request by midhun-thimmapuram**
### Benchmark comparison:

Output on Ruby 2.6.9, MacBook Pro (15-inch, 2018, 2.2 GHz 6-Core Intel Core i7)

[benchmark code](https://gist.github.com/midhun-thimmapuram/f6b45744e77f01750db64eb5f7a7a61e)

_blueprinter 0.25.3 (current master)_

```
Rehearsal -----------------------------------------------
blueprinter   0.362873   0.004129   0.367002 (  0.367198)
-------------------------------------- total: 0.367002sec

                  user     system      total        real
blueprinter   0.337027   0.001376   0.338403 (  0.338711)
Warming up --------------------------------------
         blueprinter     1.000  i/100ms
Calculating -------------------------------------
         blueprinter      2.724  (± 2.2%) i/s -     28.000  in  10.326712s
                   with 95.0% confidence
Calculating -------------------------------------
         blueprinter    63.991M memsize (     0.000  retained)
                       480.353k objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
```
_With current PR optimisations Run - 1_
```
Rehearsal -----------------------------------------------
blueprinter   0.203004   0.004435   0.207439 (  0.207799)
-------------------------------------- total: 0.207439sec

                  user     system      total        real
blueprinter   0.192406   0.000247   0.192653 (  0.192837)
Warming up --------------------------------------
         blueprinter     1.000  i/100ms
Calculating -------------------------------------
         blueprinter      4.718  (± 3.3%) i/s -     47.000  in  10.099948s
                   with 95.0% confidence
Calculating -------------------------------------
         blueprinter    16.860M memsize (     0.000  retained)
                        80.139k objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
```
_With current PR optimisations Run - 2_
```
Rehearsal -----------------------------------------------
blueprinter   0.219407   0.005306   0.224713 (  0.227365)
-------------------------------------- total: 0.224713sec

                  user     system      total        real
blueprinter   0.208257   0.000257   0.208514 (  0.208746)
Warming up --------------------------------------
         blueprinter     1.000  i/100ms
Calculating -------------------------------------
         blueprinter      4.405  (± 3.8%) i/s -     44.000  in  10.165494s
                   with 95.0% confidence
Calculating -------------------------------------
         blueprinter    16.799M memsize (     0.000  retained)
                        79.889k objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
```

### CODE REVIEW

File: lib/blueprinter/extractors/auto_extractor.rb [hunk 1]
```diff
@@ -19,7 +20,7 @@ def extract(field_name, object, local_options, options = {})
     private
 
     def default_value(field_options)
-      field_options.key?(:default) ? field_options.fetch(:default) : Blueprinter.configuration.field_default
+      field_options.key?(:default) ? field_options.fetch(:default) : @config_field_default
```
Comment (skryukov) [inline lib/blueprinter/extractors/auto_extractor.rb:23]: Have you measured this particular optimization results? `Blueprinter.configuration.field_default` seems to be just a static call 🤔

File: lib/blueprinter/formatters/date_time_formatter.rb [hunk 2]
```diff
@@ -2,8 +2,12 @@ module Blueprinter
   class DateTimeFormatter
     InvalidDateTimeFormatterError = Class.new(BlueprinterError)
 
+    def initialize
+      @config_datetime_format = Blueprinter.configuration.datetime_format
+    end
+
     def format(value, options)
-      return value if value.nil?
+      return value if value.nil? || !(options.key?(:datetime_format) || @config_datetime_format)
```
[the remark this answers is not in the document]
Comment (skryukov) [reply lib/blueprinter/formatters/date_time_formatter.rb:10]: Great! Would you mind combining changes here, so I could close my PR? 😄

[the remark this answers is not in the document]
Comment (midhun-thimmapuram) [reply lib/blueprinter/formatters/date_time_formatter.rb:10]: Sure, will do. Thank you for your review.

File: lib/blueprinter/formatters/date_time_formatter.rb [hunk 3]
```diff
@@ -3,12 +3,11 @@ class DateTimeFormatter
     InvalidDateTimeFormatterError = Class.new(BlueprinterError)
 
     def format(value, options)
-      return value if value.nil?
+      return value if value.nil? || !(field_format = options[:datetime_format] || Blueprinter.configuration.datetime_format)
```
Comment (skryukov) [inline lib/blueprinter/formatters/date_time_formatter.rb:6]: Maybe smth like this for the sake of readability?
```suggestion
      return value if value.nil?

      field_format = options[:datetime_format] || Blueprinter.configuration.datetime_format
      return value if field_format.nil?
```

**Review by skryukov — APPROVED** [review]

**Comment by midhun-thimmapuram** [comment]
@skryukov What's next for PR to be merged?

**Comment by skryukov** [comment]
We should wait for maintainers, I'm just an outsider 😅

**Review by terrence2 — APPROVED** [review]
I'm largely unfamiliar with the context here, but there appears to be two unrelated changes here.

The performance optimization is coming from the memoization of field sorting in fields_for, which is a nice, simple, and clean improvement.

There's also a mild refactor of date time formatting that appears to not change the operations that occur per format, but does clarify the code somewhat, so is a nice patch as well.

Given that these are unrelated, I'd _much_ prefer to have them land as separate patches, but not enough that I'd block getting this landed at all, given how long it's been open.

**Comment by midhun-thimmapuram** [comment]
> terrence2

Thanks for the approval, I can create another PR moving the date time formatting changes there if it is required so that we can have them as separate patches.

**Pull request closed without merging**